## Exercise 1
### Question 1
Download the dataset train.csv posted with this Exercise. This dataset is collected from Kaggle. You may also want to  
download it directly from the following Kaggle Competition (Login > Go to “Data” > “Download All” > train.csv). Either  
way, read the competition description (no login required) to get an idea about what the target Data Science task is.  
 
    a) Import the “train.csv” data you downloaded (either from NTU Learn or Kaggle) in Jupyter Notebook.
    b) How many observations (rows) and variables (columns) are in the above dataset? Check the “shape”.
    c) What are the data types (“dtypes”) – Numeric/Categorical – of the variables (columns) in the dataset?
    d) What does the .info() method do? Use the .info() method on the imported dataset to check this out.
    e) What does the .describe() method do? Use the .describe() method on the imported dataset to check.

### Import all the needed libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt # we only need pyplot
sb.set() # set the default Seaborn style for graphics

#### a) Import the “train.csv” data you downloaded (either from NTU Learn or Kaggle) in Jupyter Notebook.

Data is in csv format so pd.read_csv is used to import the data.  
Head function used to check if data was imported correctly, by checking format of first five rows.

In [2]:
traindata = pd.read_csv('train.csv')
traindata.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

#### b) How many observations (rows) and variables (columns) are in the above dataset? Check the “shape”.

To check shape, .shape() funtion is used, output is (rows,columns).

In [3]:
traindata.shape

(1460, 81)

1460 Rows, 81 columns

#### c) What are the data types (“dtypes”) – Numeric/Categorical – of the variables (columns) in the dataset?



In [4]:
traindata.sample(5)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1254  1255          60       RL         60.0     6931   Pave   NaN      Reg   
830    831          20       RL         80.0    11900   Pave   NaN      IR1   
1169  1170          60       RL        118.0    35760   Pave   NaN      IR1   
529    530          20       RL          NaN    32668   Pave   NaN      IR1   
187    188          50       RL         60.0    10410   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
1254         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
830          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1169         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
529          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
187          Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
1254      7   2008        WD         Normal     165400  
830       6   2008        WD         Normal     166000  
1169      7   2006        WD         Normal     625000  
529       3   2007        WD         Alloca     200624  
187       8   2009        WD         Normal     135000  

[5 rows x 81 columns]

Draws a random sample of five rows to see what are the values used

The data types used are a combination of numeric and categorical
| Rows | Data Type |  
|-----------|:-----------:|  
| Id | Cell A2 |  


#### d) What does the .info() method do? Use the .info() method on the imported dataset to check this out.

In [5]:
traindata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

The .info() method displays all the columns and their corresponding data types, including number of non-null entries.

#### e) What does the .describe() method do? Use the .describe() method on the imported dataset to check.

In [6]:
traindata.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

The .describe() method performs basic statistical analysis on the collumns that consist of numerical date,  
providing the count(no of data), mean(average), std deviation, min value, 25% value, 50% value, 75% value and max value.

### Question 2

Check Summer Olympic 2016 medal tally : https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table  

    a) Import the Wikipedia page in Jupyter Notebook (check M1 DataAcquisition.ipynb for hints about this).
    b) How many tables are in this Wikipedia page? Check the “len” of the imported data/page to find this out.
    c) Which one is the actual “2016 Summer Olympics medal table”? Explore all tables in the data to know.
    d) Extract the main table, “2016 Summer Olympics medal table”, and store it as a new Pandas DataFrame.
    e) Extract the TOP 20 countries from the medal table, as above, and store these rows as a new DataFrame.

### Import all the needed libraries

In [7]:
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt # we only need pyplot
sb.set() # set the default Seaborn style for graphics

#### a) Import the Wikipedia page in Jupyter Notebook (check M1 DataAcquisition.ipynb for hints about this).

Since we are importing a webpage, pd.read_html will need to be used

In [8]:
wikipage = pd.read_html('https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table')
print("Data type : ", type(wikipage))

Data type :  <class 'list'>


Data type check to find out if it is a single table or a list of tables

#### b) How many tables are in this Wikipedia page? Check the “len” of the imported data/page to find this out.

In [9]:
print("HTML tables : ", len(wikipage))

HTML tables :  7


There are 7 tables in the wikipedia page

#### c) Which one is the actual “2016 Summer Olympics medal table”? Explore all tables in the data to know.

In [10]:
wikipage

[           2016 Summer Olympics medals        2016 Summer Olympics medals.1
 0                             Location               Rio de Janeiro, Brazil
 1                           Highlights                           Highlights
 2                     Most gold medals                   United States (46)
 3                    Most total medals                  United States (121)
 4  ← 2012 Olympics medal tables 2020 →  ← 2012 Olympics medal tables 2020 →,
                                                    0
 0                                Part of a series on
 1                               2016 Summer Olympics
 2  Bid process (bid details) Development (venues,...
 3  .mw-parser-output .navbar{display:inline;font-...,
                 Rank                   NOC  Gold  Silver  Bronze  Total
 0                  1         United States    46      37      38    121
 1                  2         Great Britain    27      23      17     67
 2                  3                 China    

After reviewing the tables, the actual summer olympics medal table is the third table, so it is index 2 in the list.

In [11]:
wikipage[2]

Rank                   NOC  Gold  Silver  Bronze  Total
0                  1         United States    46      37      38    121
1                  2         Great Britain    27      23      17     67
2                  3                 China    26      18      26     70
3                  4                Russia    19      17      20     56
4                  5               Germany    17      10      15     42
..               ...                   ...   ...     ...     ...    ...
82                78               Nigeria     0       0       1      1
83                78              Portugal     0       0       1      1
84                78   Trinidad and Tobago     0       0       1      1
85                78  United Arab Emirates     0       0       1      1
86  Totals (86 NOCs)      Totals (86 NOCs)   306     307     359    972

[87 rows x 6 columns]

#### d) Extract the main table, “2016 Summer Olympics medal table”, and store it as a new Pandas DataFrame.

In [12]:
medal_table = wikipage[2]

#### e) Extract the TOP 20 countries from the medal table, as above, and store these rows as a new DataFrame.

In [13]:
top_20_countries = medal_table.head(20)
top_20_countries

Rank            NOC  Gold  Silver  Bronze  Total
0     1  United States    46      37      38    121
1     2  Great Britain    27      23      17     67
2     3          China    26      18      26     70
3     4         Russia    19      17      20     56
4     5        Germany    17      10      15     42
5     6          Japan    12       8      21     41
6     7         France    10      18      14     42
7     8    South Korea     9       3       9     21
8     9          Italy     8      12       8     28
9    10      Australia     8      11      10     29
10   11    Netherlands     8       7       4     19
11   12        Hungary     8       3       4     15
12   13        Brazil*     7       6       6     19
13   14          Spain     7       4       6     17
14   15          Kenya     6       6       1     13
15   16        Jamaica     6       3       2     11
16   17        Croatia     5       3       2     10
17   18           Cuba     5       2       4     11
18   19    New Zealand     4       9       5     18
19   20         Canada     4       3      15     22

.head() command is used to extract the first 20 results counting from the start of the dataframe  
and it is saved as a new dataframe top_20_countries.  

If only the country names and their rank are wanted, not including medal count.

In [14]:
top_20_names = top_20_countries[['Rank','NOC']]
top_20_names

Rank            NOC
0     1  United States
1     2  Great Britain
2     3          China
3     4         Russia
4     5        Germany
5     6          Japan
6     7         France
7     8    South Korea
8     9          Italy
9    10      Australia
10   11    Netherlands
11   12        Hungary
12   13        Brazil*
13   14          Spain
14   15          Kenya
15   16        Jamaica
16   17        Croatia
17   18           Cuba
18   19    New Zealand
19   20         Canada

## Bonus Problems

### A
Download the “Census Income” dataset (source : https://archive.ics.uci.edu/ml/datasets/Census+Income) from  
the UCI Machine Learning Repository (in the “Data Folder”), and import it in Jupyter Notebook as a DataFrame.  
Explore the dataset using .shape, .info() and .describe(), exactly as you did in Problem 1 above. Do you spot  
anything interesting while exploring this dataset? Discuss amongst friends or talk to the Instructor, if you did.  

In [15]:
cencus_income = pd.read_csv( 'adult.data' , sep = ',' , names = [ 'age' , 'workclass' , 'fnlwgt' , 'education' , 'education-num' , 'marital-status' ,
                                                    'occupation' , 'relationship' , 'race' , 'sex' , 'capital-gain' , 'capital-loss' , 
                                                    'hours-per-week' , 'native-country' , '>50k, <=50k'] )
cencus_income

age          workclass  fnlwgt    education  education-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

      >50k, <=50k  
0           <=50K  
1           <=50K  
2           <=50K  
3           <=50K  
4           <=50K  
...           ...  
32556       <=50K  
32557        >50K  
32558       <=50K  
32559       <=50K  
32560        >50K  

[32561 rows x 15 columns]

In [16]:
pd.read_csv( 'adult.test' , sep = ',' , names = [ 'age' , 'workclass' , 'fnlwgt' , 'education' , 'education-num' , 'marital-status' ,
                                                    'occupation' , 'relationship' , 'race' , 'sex' , 'capital-gain' , 'capital-loss' , 
                                                    'hours-per-week' , 'native-country' , '>50k, <=50k'] , skiprows = 1)

age      workclass  fnlwgt      education  education-num  \
0       25        Private  226802           11th              7   
1       38        Private   89814        HS-grad              9   
2       28      Local-gov  336951     Assoc-acdm             12   
3       44        Private  160323   Some-college             10   
4       18              ?  103497   Some-college             10   
...    ...            ...     ...            ...            ...   
16276   39        Private  215419      Bachelors             13   
16277   64              ?  321403        HS-grad              9   
16278   38        Private  374983      Bachelors             13   
16279   44        Private   83891      Bachelors             13   
16280   35   Self-emp-inc  182148      Bachelors             13   

            marital-status          occupation     relationship  \
0            Never-married   Machine-op-inspct        Own-child   
1       Married-civ-spouse     Farming-fishing          Husband   
2       Married-civ-spouse     Protective-serv          Husband   
3       Married-civ-spouse   Machine-op-inspct          Husband   
4            Never-married                   ?        Own-child   
...                    ...                 ...              ...   
16276             Divorced      Prof-specialty    Not-in-family   
16277              Widowed                   ?   Other-relative   
16278   Married-civ-spouse      Prof-specialty          Husband   
16279             Divorced        Adm-clerical        Own-child   
16280   Married-civ-spouse     Exec-managerial          Husband   

                      race      sex  capital-gain  capital-loss  \
0                    Black     Male             0             0   
1                    White     Male             0             0   
2                    White     Male             0             0   
3                    Black     Male          7688             0   
4                    White   Female             0             0   
...                    ...      ...           ...           ...   
16276                White   Female             0             0   
16277                Black     Male             0             0   
16278                White     Male             0             0   
16279   Asian-Pac-Islander     Male          5455             0   
16280                White     Male             0             0   

       hours-per-week  native-country >50k, <=50k  
0                  40   United-States      <=50K.  
1                  50   United-States      <=50K.  
2                  40   United-States       >50K.  
3                  40   United-States       >50K.  
4                  30   United-States      <=50K.  
...               ...             ...         ...  
16276              36   United-States      <=50K.  
16277              40   United-States      <=50K.  
16278              50   United-States      <=50K.  
16279              40   United-States      <=50K.  
16280              60   United-States       >50K.  

[16281 rows x 15 columns]

After reviewing the available files to download, the adult.data file has the actual census data but no names for the columns.  
After checking the adult.test file, it seeems to be a similar set but with half the data.  
The adult.names file seems to be a report on the findings of those who contributed the project and listed the column names at the end of the file.  
Will be proceeding with the full adult.data file with headers from adult.name for analysis.
fnlwgt = final weight (number of people the census believes the entry represents)
education-num = highest education in numerical form

In [17]:
cencus_income.shape

(32561, 15)

The data is a table with 32561 rows and 15 columns

In [18]:
cencus_income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  >50k, <=50k     32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


9 of the data columns are in categorical form and not numerical.

In [19]:
cencus_income.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

After reviewing the 

### B
Note that the Summer Olympic medal tally on Wikipedia follows a really nice structure for the URL, where you  
can simply change the year in https://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table to fetch any  
Summer Olympic page. Try changing 2016 in the URL to 2012 or 2008 or 2004 to see for yourself. This allows us  
to fetch the Olympics medal table from all these years (in fact, any year) quite easily. Let’s try the following.  
Write a loop to extract the main tables, “20XX Summer Olympics medal table”, from 2000 to 2016, that is, for  
the five consecutive Olympics in 2000, 2004, 2008, 2012 and 2016. Store all five tables in respective DataFrames.  
Now, extract the TOP 20 countries from each of these medal tables, and store these rows as new DataFrames.  

In [20]:
medal_tables = {}
for x in range( 2000 , 2017 , 4 ):
    table = pd.read_html(f'https://en.wikipedia.org/wiki/{x}_Summer_Olympics_medal_table')
    medal_tables[x] = table[2] #Index 2 is the position of the table we are looking for

As with the previous question, the table of medals awarded is the 3rd table and hence the index is 2.  
To ease storing of the dataframes and without needing to create multiple variables, all dataframes are saved in the  
medal_tables dictionary with the year of the competition as the key.

In [21]:
top_20 = {}
for i in medal_tables:
    top_20[i] = medal_tables[i].head(20)

The top 20 countries is following a similar concept as above with the .head() function being used and 